presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadataUpdater.java

```diff
@@ -0,0 +1,114 @@
+ * limitations under the License.
+ */
+package com.facebook.presto.hive;
+
+import com.facebook.airlift.concurrent.MoreFutures;
+import com.facebook.presto.spi.ConnectorMetadataUpdateHandle;
+import com.facebook.presto.spi.SchemaTableName;
+import com.facebook.presto.spi.connector.ConnectorMetadataUpdater;
+import com.google.common.collect.ImmutableList;
+import com.google.common.util.concurrent.SettableFuture;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Queue;
+import java.util.UUID;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.Executor;
+
+import static java.util.Objects.requireNonNull;
+
+// TODO: Revisit and make this class more robust
+public class HiveMetadataUpdater
+        implements ConnectorMetadataUpdater
+{
+    private final Executor boundedExecutor;
+
+    // Stores writerIndex <-> requestId mapping
+    private final Map<Integer, UUID> writerRequestMap = new ConcurrentHashMap<>();
+
+    // Stores requestId <-> fileNameFuture mapping
+    private final Map<UUID, SettableFuture<String>> requestFutureMap = new ConcurrentHashMap<>();
+
+    // Queue of pending requests
+    private final Queue<HiveMetadataUpdateHandle> hiveMetadataRequestQueue = new ConcurrentLinkedQueue<>();
+
+    HiveMetadataUpdater(Executor boundedExecutor)
+    {
+        this.boundedExecutor = requireNonNull(boundedExecutor, "boundedExecutor is null");
+    }
+
+    @Override
+    public List<ConnectorMetadataUpdateHandle> getPendingMetadataUpdateRequests()
+    {
+        ImmutableList.Builder<ConnectorMetadataUpdateHandle> result = ImmutableList.builder();
+        for (HiveMetadataUpdateHandle request : hiveMetadataRequestQueue) {
+            result.add(request);
+        }
+        return result.build();
+    }
+
+    @Override
+    public void setMetadataUpdateResults(List<ConnectorMetadataUpdateHandle> results)
+    {
+        boundedExecutor.execute(() -> updateResultAsync(results));
+    }
+
+    private void updateResultAsync(List<ConnectorMetadataUpdateHandle> results)
+    {
+        for (ConnectorMetadataUpdateHandle connectorMetadataUpdateHandle : results) {
+            HiveMetadataUpdateHandle updateResult = (HiveMetadataUpdateHandle) connectorMetadataUpdateHandle;
+            UUID requestId = updateResult.getRequestId();
+
+            if (!requestFutureMap.containsKey(requestId)) {
+                continue;
+            }
+
+            Optional<String> fileName = updateResult.getMetadataUpdate();
+            if (fileName.isPresent()) {
+                // remove the request from queue
+                hiveMetadataRequestQueue.removeIf(metadataUpdateRequest -> metadataUpdateRequest.getRequestId().equals(requestId));
+
+                // Set the fileName future
+                requestFutureMap.get(requestId).set(fileName.get());
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.



**Problem**  
 You are using a `ConcurrentHashMap`, but your usage of `containsKey()` and `get()` may not be thread-safe at lines: **77 and 87**. In between the check and the `get()` another thread can remove the key and the `get()` will return `null`. The remove that can remove the key is at line: **105**.

**Fix**  
 Consider calling `get()`, checking instead of your current check if the returned object is `null`, and then using that object only, without calling `get()` again.

**More info**  
 [View an example on GitHub](https://github.com/apache/hadoop/blob/f16cf877e565084c66bc63605659b157c4394dc8/hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/s3guard/S3Guard.java#L302-L304) (external link).